src/Core/Filter.re

```diff
@@ -12,38 +12,25 @@ type result('a) = {
   highlight: list((int, int)),
 };
 
-let makeResult = ((maybeMatch, item)) =>
-  switch (maybeMatch) {
-  | Some(match) => {
-      item,
-      highlight: IndexEx.ranges(match.IndexMatchResult.indicies),
-    }
-  | None => {item, highlight: []}
-  };
+let makeResult = ((item, match: Fzy.Result.t)) => {
+  item,
+  highlight: IndexEx.ranges(match.positions),
+};
 
 let rank = (query, format, items) => {
   let shouldLower = query == String.lowercase_ascii(query);
   let format = item => format(item, ~shouldLower);
 
-  let compareScore = (x, y) => {
-    let scoreObject = ((match, item)) => (
-      Option.map(m => MatchResult.create(m.IndexMatchResult.score), match),
-      format(item),
-    );
-    compareScores(scoreObject(x), scoreObject(y));
-  };
-
-  let processItem = (pattern, item) => {
-    let line = format(item);
-    let match = pathIndexMatch(~line, ~pattern);
+  let search = (query, format, items) => {
+    let searchStrings = List.map(item => format(item), items);
+    let results = Fzy.fzySearchList(searchStrings, query, ());
+    let lookup = (result: Fzy.Result.t) =>
+      List.nth(items, result.original_index);
 
-    (match, item);
+    List.map(result => (lookup(result), result), results);
```

Comment:
It looks like this logic might be O(N^2) - each of the `List.nth` calls requires a traversal of the list to pull the item out. (So for each item we `map`, we have to do a nested traversal via the `List.nth` call).

I wonder if it would be quicker to do the following:

```
let itemArray = items |> List.to_array;
let lookup = (result: Fzy.Result.t) => itemArray[result.original_index];
```

There's the extra copy of the array at the beginning (so a space trade-off).... but it means in terms of timing we wouldn't need to traverse the list for every item. If we have a benchmark in place, it would be interesting to see the comparison of the two approaches, at least.

Reply:
I'll give it a benchmark and see!

I've also realised that because this is pre-sorted, I don't need to do a final pass into `makeResult`, I can just do that inline here, removing another `List.map`.

Reply:
There isn't much in it, but it looks like `List.nth` is slightly better:

![image](https://user-images.githubusercontent.com/10038688/79040461-7ee44000-7be0-11ea-8da6-a1a1cba82075.png)

Reply:
Why is it a list in the first place? I assume it originates as an array, and I think it's going to be converted into an array a bit down the line anyway, so it might as well stay an array the whole way. Arrays can also modify items in-place, so you might save some allocation that way as well.

One of the things I noticed, and think I improved, when  I overhauled the filtering some time ago was several instances of unnecessary conversion between lists and arrays. And if given an array instead of a list it might be that more of the pipeline should work on an array. But either way, there's no need to do the list conversion prematurely, as that tends to lead to more back-and-forth, like it was before.

Reply:
In one place (Completions) it is converted back to an array so it would probably make sense there.

The other place is the `FilterJob` where I'm less sure about changing it, since the internals of the queue used there is also all list based, as well as the rest of `FilterJob`. I think you implemented the bits around there (at least the Queue), so not sure if you have any ideas if its worth refactoring it to return an array, or use an array internally (though that sounds slower).

Reply:
Right, yeah that Queue implementation definitely needs a list. I think an implementation using a list of arrays with a pointer into the "current" array could be as performant, although it would definitely be messier and would rely on the assumption that the array won't be mutated. A larger refactor of FitlerJob to be array-based might even be beneficial, although I doubt it would be more maintainable.

Whether this would be worth it depends largely on the size of the arrays I think. If they're on the order of 10 items it probably isn't. And judging by the performance of using `List.nth` they probably aren't very big.